### DISCUSSION TITLE
No username in registration flow

### DISCUSSION THREAD

**Discussion by emperorkebab** [zitadel/zitadel#8259] [General]
Ideally i would want to define the way users register and login,
the username should be either:
- a simple username
- the user email
- a username@org_domain

The problem i found is that the org domain name is weirdly enforced and can't be removed from the flow
Zitadel provides under domain settings a parameter to set this:
![image](https://github.com/zitadel/zitadel/assets/170123215/94c2192c-61a1-4497-b8d8-8801a236a752)

But there are two problems when this parameter is **disabled**:
- In the registration form the "username" field disappears (and it's not required), not letting the user chose his username
  this results in the domain name being appended by default to the username, <ins>**effectively making bare username login impossible**</ins>...
- The user login ui form placeholder still hints at providing domain even though it's technically not enforced
![image](https://github.com/zitadel/zitadel/assets/170123215/f7865617-3369-403d-903f-48e9e19aeca8)

To login with a simple username, the only solution is to login as the user (if the console is accessible) or as admin to **manually** change the username after registration, which is not applicable in production.
This very annoying as i should be able to define the login behavior, the current setup seems to only focus on multi-tenancy.

Additionally:
- when enabling then disabling that parameter all users still retain the domain in the username for some it results in a double '@' username, one with email and one with the appended domain name (upon disabling the param, the domain name should be removed)
![image](https://github.com/zitadel/zitadel/assets/170123215/e74fb285-a424-4ee4-a958-f3a89330cd88)

I took a look at the zitadel/typescript repo for using a custom ui but it's not complete and the last commit amounts back to 2 months, is it abandoned ?

**Comment by hifabienne** [comment]
Hi @emperorkebab 

If you have disabled the suffix, the user should be able to register with the email address, the email address is then taken as the username:
<img width="1045" alt="image" src="https://github.com/zitadel/zitadel/assets/38692350/c703591c-59f2-441b-8a6a-f276edee01d0">

You are right that we kind of enforce that at the moment. We also enforce some required user fields, which might not be needed in every use case.
Thats why we are working on user schemas, so you can define how your user looks like: https://github.com/zitadel/zitadel/issues/6433

Generally we have implemented a standard registration ui, if you do have different needs we recommend implementing your own registration flow.

When you disable the suffix, we do add the suffix on purpose to the usernames, as otherwise users would not be able to authenticate anymore afterwards. You can still change it via api.

The typescript repository is not yet complete, but we are working on it, we do have some package restructure ongoing right now there. but other than that, most functionality should already be implemented.

You can also choose to implement it completly on your own by using the session api

**Reply by emperorkebab** [reply]
Thank you, i'm glad the new ui repo even exists, it maybe is useful to me to tweak the registration to enforce submission of the username as defined by the user, as the new user schemas can't be used right now, should i be able to do this?

As for the suffix yeah i didn't think about mass removing the domain via api, but as an idea there could be an internal username representation that outputs the username as a combination of its value + a suffix variable would be helpful, so that the suffix can be changed without affecting it, and also ensuring the username are unique even if using a suffix, but you know better

**Reply by hifabienne** [reply]
yes you can fork the repository and change the registration form to match your needs.

the combination is basically what happens if you do set to use the domain suffix.
